### PR TITLE
ddl: support altering the other charset to utf8 or utf8mb4

### LINKED ISSUE #7920: Change default charset and collation from utf8 to utf8mb4
TiDB default charset is utf8 and collation is utf8_bin, in some cases, if the Unicode string length is more than 3 bytes(4 bytes), insert this string into a column with utf8 charset will report error like:

```
ERROR 1366 (HY000): Incorrect string value: '\xF0\xA4\x8B\xAE' for column 'v' at row 1
```

Maybe we should consider to change TiDB default charset from utf8 to utf8mb4?

@morgo what's your opinion？

### PR DESCRIPTION
**Pull request by winkyao**
### What problem does this PR solve? 

fix #7920.

Support changing the other charset to utf8 or utf8mb4. 

Rules is:

1. change any other charset to utf8mb4 is permitted, because TiDB treat all data as utf8mb4
2. change any other charset to utf8 is permitted, except utf8mb4.

Before this PR, `alter table charset` take no effects, we just parse it and ignore, in this PR, we change the charset and collate of tableInfo in the infoSchema.

Tests 

 - Unit test
 - Integration test

Code changes

 - Has exported function/method change

Side effects

 - Increased code complexity

Related changes

 - Need to update the documentation
 - Need to be included in the release note

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8037)

### CODE REVIEW

**Comment by shenli** [comment]
Please resolve the conflicts.

**Comment by winkyao** [comment]
Ci will be fixed, after the pr https://github.com/pingcap/parser/pull/14 merged.

**Comment by tiancaiamao** [comment]
Please make sure the change works
https://github.com/pingcap/parser#how-to-update-parser-for-tidb

**Comment by tiancaiamao** [comment]
PTAL @ciscoxll @crazycs520 @zimulala

File: ddl/ddl.go [hunk 1]
```diff
@@ -587,6 +588,7 @@ const (
 	codeUnsupportedShardRowIDBits    = 207
 	codeUnsupportedAddPartition      = 208
 	codeUnsupportedCoalescePartition = 209
+	codeUnsupportedModifyCharset     = 210
```
[the remark this answers is not in the document]
Comment (winkyao) [reply ddl/ddl.go]: Agree.

File: ddl/table.go [hunk 2]
```diff
@@ -379,6 +379,28 @@ func onModifyTableComment(t *meta.Meta, job *model.Job) (ver int64, _ error) {
 	return ver, nil
 }
 
+func onModifyTableCharsetAndCollate(t *meta.Meta, job *model.Job) (ver int64, _ error) {
```
Comment (ciscoxll) [inline ddl/table.go]: If there is already data in the table, there will be no problem with the direct character set ?

Comment (winkyao) [reply ddl/table.go]: Can not understand what you say.

File: ddl/ddl_api.go [hunk 3]
```diff
@@ -1035,6 +1035,17 @@ func (d *ddl) handleAutoIncID(tbInfo *model.TableInfo, schemaID int64) error {
 	return nil
 }
 
+func setDefaultTableCharsetAndCollation(tbInfo *model.TableInfo) (err error) {
+	if len(tbInfo.Charset) == 0 {
```
Comment (tiancaiamao) [inline ddl/ddl_api.go]: 1. Why check `tbInfo.Charset` and `tbInfo.Collate` in this function?
2. It never return error? then the function signature can be changed.

File: ddl/ddl_api.go [hunk 4]
```diff
@@ -1153,7 +1192,11 @@ func (d *ddl) AlterTable(ctx sessionctx.Context, ident ast.Ident, specs []*ast.A
 				case ast.TableOptionComment:
 					spec.Comment = opt.StrValue
 					err = d.AlterTableComment(ctx, ident, spec)
+				case ast.TableOptionCharset, ast.TableOptionCollate:
+					toCharset, toCollate := getCharsetAndCollateInTableOption(i, spec.Options)
```
Comment (tiancaiamao) [inline ddl/ddl_api.go]: This code is weird to me.
Why not return a charset/collate array in `getCharsetAndCollateInTableOption`, and handle the array in a loop calling `d.AlterTableCharsetAndCollate`?

File: ddl/ddl_api.go (hunk 3 above)
Comment (winkyao) [reply ddl/ddl_api.go]: 2. it return error in line 1044

Comment (winkyao) [reply ddl/ddl_api.go]: Can not understand your first question.

File: ddl/table.go (hunk 2 above)
Comment (winkyao) [reply ddl/table.go]: Because TiDB treats all data charset as utf8mb4, so change to utf8 or utf8mb4 is safe.

File: ddl/ddl_api.go (hunk 4 above)
Comment (winkyao) [reply ddl/ddl_api.go]: addressed.

**Comment by winkyao** [comment]
@tiancaiamao PTAL

**Comment by winkyao** [comment]
@zimulala PTAL

File: ddl/table.go (hunk 2 above)
Comment (ciscoxll) [reply ddl/table.go]: I understand.

**Comment by winkyao** [comment]
@zimulala PTAL

File: ddl/ddl_api.go (hunk 3 above)
Comment (tiancaiamao) [reply ddl/ddl_api.go]: From the function name `setDefaultTableCharsetAndCollation` I guess the `tbInfo.Charset` and `tbInfo.Collate` would be set unconditionally. so `if len(tbInfo.Charset)` confused me.
The real thing this function do is "setDefaultTableCharsetAndCollation" **if it's not done yet** .
That ok.

**Pull request merged**